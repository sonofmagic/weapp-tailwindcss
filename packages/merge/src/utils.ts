import type { ClassValue } from 'clsx'
import type {
  createTailwindMerge as _createTailwindMergeV3,
  extendTailwindMerge as _extendTailwindMergeV3,
  twJoin as _twJoinV3,
  twMerge as _twMergeV3,
} from 'tailwind-merge'
import type {
  createTailwindMerge as _createTailwindMergeV2,
  extendTailwindMerge as _extendTailwindMergeV2,
  twJoin as _twJoinV2,
  twMerge as _twMergeV2,
} from 'tailwind-merge-v2'
import type { CreateOptions } from './types'
import { clsx } from 'clsx'
import { escape } from 'weapp-tailwindcss/escape'

export type TailwindMergeVersion = 2 | 3
// 处理 escape 逻辑
export function noop(x: string) {
  return x
}

export interface CreateFactoryOptions<V extends TailwindMergeVersion> {
  version: V
  twMerge: V extends 3 ? typeof _twMergeV3 : typeof _twMergeV2
  twJoin: V extends 3 ? typeof _twJoinV3 : typeof _twJoinV2
  extendTailwindMerge: V extends 3 ? typeof _extendTailwindMergeV3 : typeof _extendTailwindMergeV2
  createTailwindMerge: V extends 3 ? typeof _createTailwindMergeV3 : typeof _createTailwindMergeV2
}

export function createFactory<V extends TailwindMergeVersion>(opts: CreateFactoryOptions<V>) {
  const {
    twMerge: _twMerge,
    twJoin: _twJoin,
    createTailwindMerge: _createTailwindMerge,
    extendTailwindMerge: _extendTailwindMerge,
  } = opts

  return function create(options?: CreateOptions) {
    const escapeFn = options?.escapeFn ?? escape
    const e = options?.disableEscape ? noop : escapeFn

    function twMerge(...inputs: ClassValue[]): string {
      return e(_twMerge(clsx(...inputs)))
    }

    function twJoin(...inputs: ClassValue[]): string {
      return e(_twJoin(clsx(...inputs)))
    }

    // 版本安全的 extendTailwindMerge
    const extendTailwindMerge = (...args: Parameters<typeof _extendTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _extendTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    // 版本安全的 createTailwindMerge
    const createTailwindMerge = (...args: Parameters<typeof _createTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _createTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    return {
      twMerge,
      twJoin,
      extendTailwindMerge,
      createTailwindMerge,
    }
  }
}
