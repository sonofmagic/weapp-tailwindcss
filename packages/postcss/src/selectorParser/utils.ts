// 选择器 AST 相关的通用工具，提供组合与替换的辅助函数
import type { Node } from 'postcss-selector-parser'
import type { IStyleHandlerOptions } from '../types'
import psp from 'postcss-selector-parser'

export type ParserTransformOptions = Partial<{
  lossless: boolean
  updateSelector: boolean
}>

// normalizeTransformOptions 确保 parser 在修改选择器时保持必要的副作用
export function normalizeTransformOptions(options?: ParserTransformOptions): ParserTransformOptions {
  return {
    lossless: false,
    updateSelector: true,
    ...options,
  }
}

// mklist 将节点扩展为 "节点 + combinator + 节点副本" 的结构
export function mklist(node: Node): Node[] {
  return [
    node,
    psp.combinator({
      value: '+',
    }),
    node.clone(),
  ]
}

// composeIsPseudoAst 按需把字符串数组转换为 :is(...) 结构的 AST
export function composeIsPseudoAst(strs: string | string[]): Node[] {
  if (typeof strs === 'string') {
    return mklist(psp.tag({
      value: strs,
    }))
  }
  if (strs.length > 1) {
    return mklist(psp.pseudo({
      value: ':is',
      nodes: strs.map(str =>
        psp.tag({
          value: str,
        }),
      ),
    }))
  }
  return mklist(psp.tag({
    value: strs[0],
  }))
}

// 根据配置生成替换子代选择器的 AST，默认使用 view 标签
export function getCombinatorSelectorAst(options: IStyleHandlerOptions) {
  let childCombinatorReplaceValue: Node[] = mklist(psp.tag({ value: 'view' }))
  const { cssChildCombinatorReplaceValue } = options
  if (
    typeof cssChildCombinatorReplaceValue === 'string'
    || (Array.isArray(cssChildCombinatorReplaceValue) && cssChildCombinatorReplaceValue.length > 0)) {
    childCombinatorReplaceValue = composeIsPseudoAst(cssChildCombinatorReplaceValue)
  }
  return childCombinatorReplaceValue
}
