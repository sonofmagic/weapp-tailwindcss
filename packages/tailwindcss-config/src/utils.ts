export {
  defu,
  defuOverrideArray,
} from '@weapp-tailwindcss/shared'
