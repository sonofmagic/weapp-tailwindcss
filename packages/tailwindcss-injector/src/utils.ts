export {
  defu,
  defuOverrideArray,
  isRegexp,
  regExpTest,
} from '@weapp-tailwindcss/shared'

export {
  md5,
  removeAllExtensions as removeFileExtension,
} from '@weapp-tailwindcss/shared/node'
