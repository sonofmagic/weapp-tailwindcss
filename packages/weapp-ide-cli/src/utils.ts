import process from 'node:process'
import path from 'pathe'
import { execa } from 'execa'
import type { AliasEntry } from './types'

export async function execute(cliPath: string, argv: string[]) {
  const task = execa(cliPath, argv)

  task?.stdout?.pipe(process.stdout)

  await task
  // 调用返回码为 0 时代表正常，为 -1 时错误。
  // task.exitCode === 0
  // task.exitCode === -1
}

export function resolvePath(filePath: string) {
  if (path.isAbsolute(filePath)) {
    return filePath
  }
  else {
    return path.resolve(process.cwd(), filePath)
  }
}

function alias(argv: string[], entry: AliasEntry) {
  let findIdx = argv.indexOf(entry.find)
  // alias -p as --project
  if (findIdx > -1) {
    argv[findIdx] = entry.replacement
  }
  else {
    findIdx = argv.indexOf(entry.replacement)
  }

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

function pathCompat(argv: string[], option: string) {
  const findIdx = argv.indexOf(option)

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

export function createAlias(entry: AliasEntry) {
  return function (argv: string[]) {
    return alias(argv, entry)
  }
}

export function createPathCompat(option: string) {
  return function (argv: string[]) {
    return pathCompat(argv, option)
  }
}
