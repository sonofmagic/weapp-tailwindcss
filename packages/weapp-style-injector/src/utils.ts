import type { PerFileImportResolver } from './core'

export function toArray<T>(value: T | T[] | null | undefined): Array<NonNullable<T>> {
  if (value == null) {
    return []
  }
  return (Array.isArray(value) ? value : [value]) as Array<NonNullable<T>>
}

export function mergePerFileResolvers(
  resolvers: Array<PerFileImportResolver | null | undefined>,
): PerFileImportResolver | undefined {
  const activeResolvers = resolvers.filter((resolver): resolver is PerFileImportResolver => typeof resolver === 'function')

  if (activeResolvers.length === 0) {
    return undefined
  }

  return (fileName: string) => activeResolvers.flatMap(resolver => toArray(resolver(fileName)))
}

export function ensurePosix(value: string): string {
  return value.replace(/\\/g, '/')
}

export function normalizeRoot(root: string): string {
  const trimmed = root.trim().replace(/^[./\\]+/, '').replace(/\\+/g, '/')
  return trimmed.endsWith('/') ? trimmed.slice(0, -1) : trimmed
}

export function normalizeRelativeImport(target: string): string {
  if (target.startsWith('.') || target.startsWith('/')) {
    return target
  }
  return `./${target}`
}
