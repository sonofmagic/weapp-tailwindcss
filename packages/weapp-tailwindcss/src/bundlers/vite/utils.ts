import type { ExistingRawSourceMap } from 'rollup'
import path from 'node:path'
import process from 'node:process'

export function slash(p: string): string {
  return p.replace(/\\/g, '/')
}

export const isWindows = process.platform === 'win32'

const cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`
export const cssLangRE = new RegExp(cssLangs)
export function isCSSRequest(request: string): boolean {
  return cssLangRE.test(request)
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

const postfixRE = /[?#].*$/
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '')
}

export async function formatPostcssSourceMap(
  rawMap: ExistingRawSourceMap,
  file: string,
): Promise<ExistingRawSourceMap> {
  const inputFileDir = path.dirname(file)

  const sources = rawMap.sources.map((source) => {
    const cleanSource = cleanUrl(decodeURIComponent(source))

    // postcss virtual files
    if (cleanSource[0] === '<' && cleanSource.endsWith('>')) {
      return `\0${cleanSource}`
    }

    return normalizePath(path.resolve(inputFileDir, cleanSource))
  })

  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version,
  }
}
