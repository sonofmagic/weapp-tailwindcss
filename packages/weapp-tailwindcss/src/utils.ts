import { createDefu } from 'defu'
// import type { Defu } from 'defu'
import type { InternalUserDefinedOptions } from '@/types'

export function isRegexp(value: unknown) {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isMap(value: unknown) {
  return Object.prototype.toString.call(value) === '[object Map]'
}
export function regExpTest(arr: (string | RegExp)[] = [], str: string) {
  if (Array.isArray(arr)) {
    for (const item of arr) {
      if (typeof item === 'string') {
        if (str.includes(item)) {
          return true
        }
      }
      else if (isRegexp(item)) {
        item.lastIndex = 0
        if (item.test(str)) {
          return true
        }
      }
    }
    return false
  }
  throw new TypeError('paramater \'arr\' should be a Array of Regexp | String !')
}

export function noop() {}

// const MAX_ASCII_CHAR_CODE = 127

// export function isAscii(str: string) {
//   for (let i = 0, strLen = str.length; i < strLen; ++i) {
//     if (str.charCodeAt(i) > MAX_ASCII_CHAR_CODE) { return false }
//   }
//   return true
// }

export function groupBy<T>(arr: T[], cb: (arg: T) => string): Record<string, T[]> {
  if (!Array.isArray(arr)) {
    throw new TypeError('expected an array for first argument')
  }

  if (typeof cb !== 'function') {
    throw new TypeError('expected a function for second argument')
  }

  const result: Record<string, T[]> = {}
  for (const item of arr) {
    const bucketCategory = cb(item)
    const bucket = result[bucketCategory]

    if (Array.isArray(bucket)) {
      result[bucketCategory].push(item)
    }
    else {
      result[bucketCategory] = [item]
    }
  }

  return result
}

export function getGroupedEntries<T>(entries: [string, T][], options: InternalUserDefinedOptions) {
  const { cssMatcher, htmlMatcher, jsMatcher, wxsMatcher } = options
  const groupedEntries = groupBy(entries, ([file]) => {
    if (cssMatcher(file)) {
      return 'css'
    }
    else if (htmlMatcher(file)) {
      return 'html'
    }
    else if (jsMatcher(file) || wxsMatcher(file)) {
      return 'js'
    }
    else {
      return 'other'
    }
  })
  return groupedEntries as Record<'css' | 'html' | 'js' | 'other', [string, T][]>
}

export function removeExt(file: string) {
  return file.replace(/\.[^./]+$/, '')
}

export { default as defu } from 'defu'

export const defuOverrideArray = createDefu((obj, key, value) => {
  if (Array.isArray(obj[key]) && Array.isArray(value)) {
    obj[key] = value
    return true
  }
})
