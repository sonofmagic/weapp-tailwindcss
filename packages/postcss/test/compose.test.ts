import fs from 'fs-extra'
import path from 'pathe'
import { createStyleHandler } from '@/index'
import { generateCss } from './utils'

async function compose(base: string) {
  const styleHandler = createStyleHandler({
    isMainChunk: true,
  })
  const { css: code } = await generateCss(
    path.resolve(__dirname, base),
    {
      css: await fs.readFile(path.resolve(__dirname, base, './index.css'), 'utf8'),
    },
  )
  expect(code).toMatchSnapshot()
  const { css } = await styleHandler(code)
  expect(css).toMatchSnapshot()
}

describe('compose', () => {
  it('components 0', async () => {
    await compose('./fixtures/css/compose-0')
  })

  it('no components 0', async () => {
    await compose('./fixtures/css/compose-1')
  })
})
