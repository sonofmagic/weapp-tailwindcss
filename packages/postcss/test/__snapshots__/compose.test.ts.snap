// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compose > components 0 1`] = `
"/*! tailwindcss v4.1.11 | MIT License | https://tailwindcss.com */
@layer properties;
:root, :host {
  --font-sans: ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  --color-white: #fff;
  --spacing: .25rem;
  --font-weight-medium: 500;
  --radius-sm: .25rem;
}
*,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
@layer components {
  .btn-primary {
    border-radius: var(--radius-sm);
    padding-block: calc(var(--spacing) * 4);
    text-align: center;
    font-family: var(--font-sans);
    --tw-font-weight: var(--font-weight-medium);
    font-weight: var(--font-weight-medium);
    color: var(--color-white);
  }
}
@property --tw-font-weight {
  syntax: "*";
  inherits: false;
}
@layer properties {
  @supports ((-webkit-hyphens: none) and (not (margin-trim: inline))) or ((-moz-orient: inline) and (not (color:rgb(from red r g b)))) {
    *, ::before, ::after, ::backdrop {
      --tw-font-weight: initial;
    }
  }
}
"
`;

exports[`compose > components 0 2`] = `
"/*! tailwindcss v4.1.11 | MIT License | https://tailwindcss.com */
view,text,::before,::after,::backdrop {
      --tw-font-weight: initial;
    }
page,:host {
  --font-sans: ui-sans-serif,system-ui, -apple-system, Segoe UI, Roboto, Ubuntu, Cantarell, Noto Sans, sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  --color-white: #fff;
  --spacing: .25rem;
  --font-weight-medium: 500;
  --radius-sm: .25rem;
}
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.btn-primary {
    border-radius: var(--radius-sm);
    padding-top: calc(var(--spacing) * 4);
    padding-bottom: calc(var(--spacing) * 4);
    text-align: center;
    font-family: var(--font-sans);
    --tw-font-weight: var(--font-weight-medium);
    font-weight: var(--font-weight-medium);
    color: var(--color-white);
  }
"
`;

exports[`compose > no components 0 1`] = `
"/*! tailwindcss v4.1.11 | MIT License | https://tailwindcss.com */
@layer properties;
:root, :host {
  --font-sans: ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  --color-white: #fff;
  --spacing: .25rem;
  --font-weight-medium: 500;
  --radius-sm: .25rem;
}
*,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.btn-primary {
  border-radius: var(--radius-sm);
  padding-block: calc(var(--spacing) * 4);
  text-align: center;
  font-family: var(--font-sans);
  --tw-font-weight: var(--font-weight-medium);
  font-weight: var(--font-weight-medium);
  color: var(--color-white);
}
@property --tw-font-weight {
  syntax: "*";
  inherits: false;
}
@layer properties {
  @supports ((-webkit-hyphens: none) and (not (margin-trim: inline))) or ((-moz-orient: inline) and (not (color:rgb(from red r g b)))) {
    *, ::before, ::after, ::backdrop {
      --tw-font-weight: initial;
    }
  }
}
"
`;

exports[`compose > no components 0 2`] = `
"/*! tailwindcss v4.1.11 | MIT License | https://tailwindcss.com */
view,text,::before,::after,::backdrop {
      --tw-font-weight: initial;
    }
page,:host {
  --font-sans: ui-sans-serif,system-ui, -apple-system, Segoe UI, Roboto, Ubuntu, Cantarell, Noto Sans, sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  --color-white: #fff;
  --spacing: .25rem;
  --font-weight-medium: 500;
  --radius-sm: .25rem;
}
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.btn-primary {
  border-radius: var(--radius-sm);
  padding-top: calc(var(--spacing) * 4);
  padding-bottom: calc(var(--spacing) * 4);
  text-align: center;
  font-family: var(--font-sans);
  --tw-font-weight: var(--font-weight-medium);
  font-weight: var(--font-weight-medium);
  color: var(--color-white);
}
"
`;
